yourbench/utils/dataset_engine.py

```
@@ -8,342 +10,273 @@
+
+
+def _merge_with_existing(
+    existing: Dataset | DatasetDict, new: Dataset, subset: Optional[str]
+) -> Dataset | DatasetDict:
+    """Create a fresh Dataset/DatasetDict with *new* inserted (no in‑place mutation)."""
+
+    if subset is None:
+        return new
+
+    if not isinstance(existing, DatasetDict):
+        existing = DatasetDict({"default": existing})
+
+    merged = DatasetDict({k: v for k, v in existing.items() if k != subset})
+    merged[subset] = new
+    return merged
+
+
+def _safe_save_to_disk(dataset: Dataset | DatasetDict, path: Path) -> None:
+    """Handle the dreaded “dataset can't overwrite itself” error transparently."""
+
+    try:
+        dataset.save_to_disk(str(path))
+        return
+    except PermissionError as exc:
+        if "dataset can't overwrite itself" not in str(exc):
+            raise
+
+    import shutil
+    import tempfile
+
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp_path = Path(tmp)
+        dataset.save_to_disk(tmp)
+        shutil.rmtree(path, ignore_errors=True)
+        shutil.copytree(tmp_path, path)
+
+    logger.success("Dataset saved via temporary directory to avoid overwrite issue")
+
+
+# ---------------------------------------------------------------------------
+# Public API (stable signatures)
+# ---------------------------------------------------------------------------
+
+
+def custom_load_dataset(config: Dict[str, Any], subset: Optional[str] = None) -> Dataset:  # noqa: D401
+    """Return *subset* of the dataset described by *config* – local first, then Hub."""
+
+    settings = _extract_hf_settings(config)
+
+    if settings.local_dir and settings.local_dir.exists():
+        logger.info(f"Loading dataset from '{settings.local_dir}'")
+        return _load_from_local(settings.local_dir, subset)
+
+    if _is_offline_mode():
+        logger.warning("Offline mode with no local data; returning empty Dataset")
         return Dataset.from_dict({})
```

Returning an empty dataset in offline mode hides errors. Suggest raising:
```
raise RuntimeError("Offline mode is enabled but no local dataset was found. Cannot proceed.")
```